File: web_graphing_tool/parser.js
/**
 * Mathematical Expression Parser
 * Implements a Pratt Parser for robust expression parsing
 */

class Parser {
    constructor() {
        this.functions = {
            'sin': Math.sin,
            'cos': Math.cos,
            'tan': Math.tan,
            'asin': Math.asin,
            'acos': Math.acos,
            'atan': Math.atan,
            'sinh': Math.sinh,
            'cosh': Math.cosh,
            'tanh': Math.tanh,
            'log': Math.log10,
            'ln': Math.log,
            'exp': Math.exp,
            'sqrt': Math.sqrt,
            'abs': Math.abs,
            'floor': Math.floor,
            'ceil': Math.ceil,
            'round': Math.round,
            'sign': Math.sign,
            'min': Math.min,
            'max': Math.max
        };
        
        this.constants = {
            'pi': Math.PI,
            'e': Math.E,
            'phi': (1 + Math.sqrt(5)) / 2
        };
    }

    /**
     * Tokenize the input string
     */
    tokenize(expr) {
        const tokens = [];
        let i = 0;
        const str = expr.replace(/\s+/g, '');
        
        while (i < str.length) {
            const char = str[i];
            
            // Numbers (including decimals and scientific notation)
            if (/[0-9]/.test(char) || (char === '.' && /[0-9]/.test(str[i + 1]))) {
                let num = '';
                while (i < str.length && /[0-9.eE+-]/.test(str[i])) {
                    // Handle scientific notation
                    if ((str[i] === 'e' || str[i] === 'E') && i + 1 < str.length) {
                        if (/[0-9]/.test(str[i - 1]) && (/[0-9]/.test(str[i + 1]) || 
                            ((str[i + 1] === '+' || str[i + 1] === '-') && /[0-9]/.test(str[i + 2])))) {
                            num += str[i++];
                            continue;
                        }
                    }
                    if (str[i] === '+' || str[i] === '-') {
                        if (num.includes('e') || num.includes('E')) {
                            num += str[i++];
                            continue;
                        }
                    }
                    if (str[i] === '.' && num.includes('.')) break;
                    num += str[i++];
                }
                tokens.push({ type: 'NUMBER', value: parseFloat(num) });
            }
            // Identifiers (functions, variables, constants)
            else if (/[a-zA-Z_]/.test(char)) {
                let id = '';
                while (i < str.length && /[a-zA-Z0-9_]/.test(str[i])) {
                    id += str[i++];
                }
                if (this.functions[id.toLowerCase()]) {
                    tokens.push({ type: 'FUNCTION', value: id.toLowerCase() });
                } else if (this.constants[id.toLowerCase()]) {
                    tokens.push({ type: 'CONSTANT', value: this.constants[id.toLowerCase()] });
                } else {
                    tokens.push({ type: 'VARIABLE', value: id });
                }
            }
            // Operators
            else if ('+-*/^'.includes(char)) {
                tokens.push({ type: 'OPERATOR', value: char });
                i++;
            }
            // Parentheses
            else if (char === '(') {
                tokens.push({ type: 'LPAREN', value: '(' });
                i++;
            }
            else if (char === ')') {
                tokens.push({ type: 'RPAREN', value: ')' });
                i++;
            }
            // Comma (for multi-argument functions)
            else if (char === ',') {
                tokens.push({ type: 'COMMA', value: ',' });
                i++;
            }
            else {
                throw new Error(`Unexpected character: ${char}`);
            }
        }
        
        return tokens;
    }

    /**
     * Parse tokens into an AST using Pratt Parser
     */
    parse(tokens) {
        this.tokens = tokens;
        this.pos = 0;
        
        const ast = this.parseExpression();
        
        if (this.pos < this.tokens.length) {
            throw new Error(`Unexpected token: ${this.tokens[this.pos].value}`);
        }
        
        return ast;
    }

    parseExpression() {
        return this.parseAdditive();
    }

    parseAdditive() {
        let left = this.parseMultiplicative();
        
        while (this.peek() && this.peek().type === 'OPERATOR' && 
               (this.peek().value === '+' || this.peek().value === '-')) {
            const op = this.consume().value;
            const right = this.parseMultiplicative();
            left = { type: 'BINARY_OP', operator: op, left, right };
        }
        
        return left;
    }

    parseMultiplicative() {
        let left = this.parsePower();
        
        while (this.peek() && this.peek().type === 'OPERATOR' && 
               (this.peek().value === '*' || this.peek().value === '/')) {
            const op = this.consume().value;
            const right = this.parsePower();
            left = { type: 'BINARY_OP', operator: op, left, right };
        }
        
        return left;
    }

    parsePower() {
        let left = this.parseUnary();
        
        while (this.peek() && this.peek().type === 'OPERATOR' && this.peek().value === '^') {
            const op = this.consume().value;
            const right = this.parseUnary();
            left = { type: 'BINARY_OP', operator: op, left, right };
        }
        
        return left;
    }

    parseUnary() {
        if (this.peek() && this.peek().type === 'OPERATOR' && 
            (this.peek().value === '-' || this.peek().value === '+')) {
            const op = this.consume().value;
            const operand = this.parseUnary();
            if (op === '-') {
                return { type: 'UNARY_OP', operator: '-', operand };
            }
            return operand;
        }
        return this.parsePrimary();
    }

    parsePrimary() {
        const token = this.peek();
        
        if (!token) {
            throw new Error('Unexpected end of expression');
        }
        
        switch (token.type) {
            case 'NUMBER':
                this.consume();
                return { type: 'NUMBER', value: token.value };
                
            case 'CONSTANT':
                this.consume();
                return { type: 'NUMBER', value: token.value };
                
            case 'VARIABLE':
                this.consume();
                return { type: 'VARIABLE', name: token.value };
                
            case 'FUNCTION':
                this.consume();
                if (!this.peek() || this.peek().type !== 'LPAREN') {
                    throw new Error('Expected opening parenthesis after function');
                }
                this.consume(); // consume '('
                
                const args = [this.parseExpression()];
                while (this.peek() && this.peek().type === 'COMMA') {
                    this.consume(); // consume ','
                    args.push(this.parseExpression());
                }
                
                if (!this.peek() || this.peek().type !== 'RPAREN') {
                    throw new Error('Expected closing parenthesis');
                }
                this.consume(); // consume ')'
                
                return { type: 'FUNCTION_CALL', name: token.value, args };
                
            case 'LPAREN':
                this.consume();
                const expr = this.parseExpression();
                if (!this.peek() || this.peek().type !== 'RPAREN') {
                    throw new Error('Expected closing parenthesis');
                }
                this.consume();
                return expr;
                
            default:
                throw new Error(`Unexpected token: ${token.value}`);
        }
    }

    peek() {
        return this.tokens[this.pos];
    }

    consume() {
        return this.tokens[this.pos++];
    }

    /**
     * Compile AST to an evaluatable function
     */
    compile(ast, variableName = 'x') {
        const code = this.generateCode(ast);
        
        // Create a safe evaluation function
        const funcBody = `
            const sin = Math.sin;
            const cos = Math.cos;
            const tan = Math.tan;
            const asin = Math.asin;
            const acos = Math.acos;
            const atan = Math.atan;
            const sinh = Math.sinh;
            const cosh = Math.cosh;
            const tanh = Math.tanh;
            const log = Math.log10;
            const ln = Math.log;
            const exp = Math.exp;
            const sqrt = Math.sqrt;
            const abs = Math.abs;
            const floor = Math.floor;
            const ceil = Math.ceil;
            const round = Math.round;
            const sign = Math.sign;
            const min = Math.min;
            const max = Math.max;
            const pi = ${Math.PI};
            const e = ${Math.E};
            const phi = ${(1 + Math.sqrt(5)) / 2};
            return (${code});
        `;
        
        try {
            const fn = new Function(variableName, funcBody);
            return { valid: true, fn };
        } catch (e) {
            return { valid: false, error: e.message };
        }
    }

    generateCode(node) {
        if (!node) return '';
        
        switch (node.type) {
            case 'NUMBER':
                return node.value.toString();
                
            case 'VARIABLE':
                return node.name;
                
            case 'UNARY_OP':
                return `(-${this.generateCode(node.operand)})`;
                
            case 'BINARY_OP':
                const left = this.generateCode(node.left);
                const right = this.generateCode(node.right);
                if (node.operator === '^') {
                    return `Math.pow(${left}, ${right})`;
                }
                return `(${left} ${node.operator} ${right})`;
                
            case 'FUNCTION_CALL':
                const args = node.args.map(arg => this.generateCode(arg)).join(', ');
                return `${node.name}(${args})`;
                
            default:
                throw new Error(`Unknown node type: ${node.type}`);
        }
    }

    /**
     * Parse and compile in one step
     */
    parseAndCompile(expr, variableName = 'x') {
        try {
            const tokens = this.tokenize(expr);
            const ast = this.parse(tokens);
            return this.compile(ast, variableName);
        } catch (e) {
            return { valid: false, error: e.message };
        }
    }
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = Parser;
}

File: graphing_tool/src/parser.py
"""
Mathematical Expression Parser and AST Engine

This module implements a complete lexer, parser (using Pratt parsing), 
and evaluator for mathematical expressions.
"""

import math
import re
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple, Union
from enum import Enum, auto


class TokenType(Enum):
    """Enumeration of token types for lexical analysis."""
    NUMBER = auto()
    IDENTIFIER = auto()
    PLUS = auto()
    MINUS = auto()
    MULTIPLY = auto()
    DIVIDE = auto()
    POWER = auto()
    LPAREN = auto()
    RPAREN = auto()
    COMMA = auto()
    EOF = auto()


@dataclass
class Token:
    """Represents a single token from lexical analysis."""
    type: TokenType
    value: Any
    position: int


class Lexer:
    """
    Tokenizes mathematical expression strings into structured tokens.
    
    Supports:
    - Numbers (integers and floats)
    - Identifiers (variables and function names)
    - Operators: +, -, *, /, ^
    - Delimiters: (, ), ,
    - Constants: pi, e
    """
    
    # Built-in constants
    CONSTANTS = {
        'pi': math.pi,
        'e': math.e,
        'τ': math.tau,
    }
    
    # Built-in functions
    FUNCTIONS = {
        'sin': math.sin,
        'cos': math.cos,
        'tan': math.tan,
        'asin': math.asin,
        'acos': math.acos,
        'atan': math.atan,
        'sinh': math.sinh,
        'cosh': math.cosh,
        'tanh': math.tanh,
        'log': math.log10,
        'ln': math.log,
        'log2': math.log2,
        'exp': math.exp,
        'sqrt': math.sqrt,
        'abs': abs,
        'floor': math.floor,
        'ceil': math.ceil,
        'round': round,
        'sign': lambda x: (x > 0) - (x < 0),
        'max': max,
        'min': min,
    }
    
    def __init__(self, text: str):
        self.text = text
        self.pos = 0
        self.length = len(text)
        
    def _skip_whitespace(self):
        """Skip whitespace characters."""
        while self.pos < self.length and self.text[self.pos].isspace():
            self.pos += 1
    
    def _read_number(self) -> Token:
        """Read a numeric literal (integer or float)."""
        start = self.pos
        
        # Handle negative numbers at the start or after an operator
        if self.pos < self.length and self.text[self.pos] == '-':
            self.pos += 1
        
        # Read integer part
        while self.pos < self.length and self.text[self.pos].isdigit():
            self.pos += 1
        
        # Read decimal part
        if self.pos < self.length and self.text[self.pos] == '.':
            self.pos += 1
            while self.pos < self.length and self.text[self.pos].isdigit():
                self.pos += 1
        
        # Read exponent part
        if self.pos < self.length and self.text[self.pos].lower() == 'e':
            exp_pos = self.pos
            self.pos += 1
            if self.pos < self.length and self.text[self.pos] in '+-':
                self.pos += 1
            if self.pos < self.length and self.text[self.pos].isdigit():
                while self.pos < self.length and self.text[self.pos].isdigit():
                    self.pos += 1
            else:
                self.pos = exp_pos  # Not a valid exponent, backtrack
        
        num_str = self.text[start:self.pos]
        return Token(TokenType.NUMBER, float(num_str), start)
    
    def _read_identifier(self) -> Token:
        """Read an identifier (variable name or function name)."""
        start = self.pos
        
        while self.pos < self.length and (
            self.text[self.pos].isalnum() or self.text[self.pos] == '_'
        ):
            self.pos += 1
        
        value = self.text[start:self.pos]
        return Token(TokenType.IDENTIFIER, value, start)
    
    def tokenize(self) -> List[Token]:
        """Convert the input string into a list of tokens."""
        tokens = []
        
        while self.pos < self.length:
            self._skip_whitespace()
            
            if self.pos >= self.length:
                break
            
            char = self.text[self.pos]
            start_pos = self.pos
            
            if char.isdigit() or (char == '.' and self.pos + 1 < self.length and self.text[self.pos + 1].isdigit()):
                tokens.append(self._read_number())
            elif char.isalpha() or char == '_':
                tokens.append(self._read_identifier())
            elif char == '+':
                tokens.append(Token(TokenType.PLUS, '+', start_pos))
                self.pos += 1
            elif char == '-':
                tokens.append(Token(TokenType.MINUS, '-', start_pos))
                self.pos += 1
            elif char == '*':
                tokens.append(Token(TokenType.MULTIPLY, '*', start_pos))
                self.pos += 1
            elif char == '/':
                tokens.append(Token(TokenType.DIVIDE, '/', start_pos))
                self.pos += 1
            elif char == '^':
                tokens.append(Token(TokenType.POWER, '^', start_pos))
                self.pos += 1
            elif char == '(':
                tokens.append(Token(TokenType.LPAREN, '(', start_pos))
                self.pos += 1
            elif char == ')':
                tokens.append(Token(TokenType.RPAREN, ')', start_pos))
                self.pos += 1
            elif char == ',':
                tokens.append(Token(TokenType.COMMA, ',', start_pos))
                self.pos += 1
            else:
                raise SyntaxError(f"Unexpected character '{char}' at position {start_pos}")
        
        tokens.append(Token(TokenType.EOF, None, self.pos))
        return tokens


# AST Node Types
@dataclass
class ASTNode:
    """Base class for AST nodes."""
    pass


@dataclass
class NumberNode(ASTNode):
    """AST node for numeric literals."""
    value: float


@dataclass
class VariableNode(ASTNode):
    """AST node for variables."""
    name: str


@dataclass
class BinaryOpNode(ASTNode):
    """AST node for binary operations."""
    op: str
    left: ASTNode
    right: ASTNode


@dataclass
class UnaryOpNode(ASTNode):
    """AST node for unary operations."""
    op: str
    operand: ASTNode


@dataclass
class FunctionCallNode(ASTNode):
    """AST node for function calls."""
    name: str
    args: List[ASTNode]


class Parser:
    """
    Pratt parser for converting tokens into an Abstract Syntax Tree.
    
    Implements operator precedence parsing with support for:
    - Binary operators: +, -, *, /, ^
    - Unary operators: +, -
    - Function calls
    - Parenthesized expressions
    """
    
    # Operator precedence (higher = binds tighter)
    PRECEDENCE = {
        '+': 10,
        '-': 10,
        '*': 20,
        '/': 20,
        '^': 30,
    }
    
    # Right-associative operators
    RIGHT_ASSOC = {'^'}
    
    def __init__(self, tokens: List[Token]):
        self.tokens = tokens
        self.pos = 0
        
    def _current_token(self) -> Token:
        """Get the current token."""
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return self.tokens[-1]  # EOF
    
    def _consume(self, expected_type: TokenType = None) -> Token:
        """Consume and return the current token."""
        token = self._current_token()
        if expected_type and token.type != expected_type:
            raise SyntaxError(
                f"Expected {expected_type}, got {token.type} at position {token.position}"
            )
        self.pos += 1
        return token
    
    def _parse_expression(self, min_precedence: int = 0) -> ASTNode:
        """Parse an expression using Pratt parsing algorithm."""
        node = self._parse_prefix()
        
        while True:
            token = self._current_token()
            
            if token.type == TokenType.EOF:
                break
            
            if token.type not in (TokenType.PLUS, TokenType.MINUS, 
                                   TokenType.MULTIPLY, TokenType.DIVIDE, 
                                   TokenType.POWER):
                break
            
            op = token.value
            prec = self.PRECEDENCE.get(op, 0)
            
            if prec < min_precedence:
                break
            
            # Handle right-associativity
            if op in self.RIGHT_ASSOC:
                next_prec = prec
            else:
                next_prec = prec + 1
            
            self._consume()
            right = self._parse_expression(next_prec)
            node = BinaryOpNode(op, node, right)
        
        return node
    
    def _parse_prefix(self) -> ASTNode:
        """Parse prefix expressions (unary operators, numbers, identifiers)."""
        token = self._current_token()
        
        if token.type == TokenType.NUMBER:
            self._consume()
            return NumberNode(token.value)
        
        if token.type == TokenType.IDENTIFIER:
            name = token.value
            self._consume()
            
            # Check if it's a function call
            if self._current_token().type == TokenType.LPAREN:
                self._consume(TokenType.LPAREN)
                args = []
                
                if self._current_token().type != TokenType.RPAREN:
                    args.append(self._parse_expression())
                    
                    while self._current_token().type == TokenType.COMMA:
                        self._consume(TokenType.COMMA)
                        args.append(self._parse_expression())
                
                self._consume(TokenType.RPAREN)
                return FunctionCallNode(name, args)
            
            # Check if it's a constant
            if name in Lexer.CONSTANTS:
                return NumberNode(Lexer.CONSTANTS[name])
            
            # It's a variable
            return VariableNode(name)
        
        if token.type == TokenType.PLUS:
            self._consume()
            operand = self._parse_prefix()
            return UnaryOpNode('+', operand)
        
        if token.type == TokenType.MINUS:
            self._consume()
            operand = self._parse_prefix()
            return UnaryOpNode('-', operand)
        
        if token.type == TokenType.LPAREN:
            self._consume(TokenType.LPAREN)
            node = self._parse_expression()
            self._consume(TokenType.RPAREN)
            return node
        
        raise SyntaxError(f"Unexpected token {token.type} at position {token.position}")
    
    def parse(self) -> ASTNode:
        """Parse the token stream into an AST."""
        ast = self._parse_expression()
        
        if self._current_token().type != TokenType.EOF:
            raise SyntaxError(
                f"Unexpected token {self._current_token().type} at position {self._current_token().position}"
            )
        
        return ast


class Evaluator:
    """
    Evaluates an AST with given variable bindings.
    
    Provides fast numerical evaluation by traversing the pre-built AST.
    """
    
    def __init__(self):
        self.functions = Lexer.FUNCTIONS.copy()
        self.constants = Lexer.CONSTANTS.copy()
    
    def evaluate(self, node: ASTNode, variables: Dict[str, float] = None) -> float:
        """Evaluate an AST node with the given variable bindings."""
        variables = variables or {}
        
        if isinstance(node, NumberNode):
            return node.value
        
        if isinstance(node, VariableNode):
            if node.name in variables:
                return variables[node.name]
            if node.name in self.constants:
                return self.constants[node.name]
            raise NameError(f"Undefined variable: {node.name}")
        
        if isinstance(node, UnaryOpNode):
            operand = self.evaluate(node.operand, variables)
            if node.op == '+':
                return operand
            if node.op == '-':
                return -operand
            raise ValueError(f"Unknown unary operator: {node.op}")
        
        if isinstance(node, BinaryOpNode):
            left = self.evaluate(node.left, variables)
            right = self.evaluate(node.right, variables)
            
            if node.op == '+':
                return left + right
            if node.op == '-':
                return left - right
            if node.op == '*':
                return left * right
            if node.op == '/':
                if right == 0:
                    raise ZeroDivisionError("Division by zero")
                return left / right
            if node.op == '^':
                try:
                    result = left ** right
                    if isinstance(result, complex):
                        raise ValueError("Result is complex")
                    return result
                except (ValueError, OverflowError):
                    raise ValueError(f"Invalid power operation: {left}^{right}")
            
            raise ValueError(f"Unknown binary operator: {node.op}")
        
        if isinstance(node, FunctionCallNode):
            func_name = node.name.lower()
            
            if func_name not in self.functions:
                raise NameError(f"Undefined function: {func_name}")
            
            func = self.functions[func_name]
            args = [self.evaluate(arg, variables) for arg in node.args]
            
            try:
                return func(*args)
            except (ValueError, ZeroDivisionError, OverflowError) as e:
                raise ValueError(f"Function error in {func_name}: {e}")
        
        raise TypeError(f"Unknown node type: {type(node)}")


class CompiledExpression:
    """
    A compiled mathematical expression ready for fast evaluation.
    
    Combines lexing, parsing, and provides optimized evaluation methods.
    """
    
    def __init__(self, expression: str, default_vars: Dict[str, float] = None):
        self.expression = expression
        self.default_vars = default_vars or {}
        
        # Compile the expression
        lexer = Lexer(expression)
        tokens = lexer.tokenize()
        parser = Parser(tokens)
        self.ast = parser.parse()
        
        self.evaluator = Evaluator()
    
    def evaluate(self, variables: Dict[str, float] = None) -> float:
        """Evaluate the expression with optional variable overrides."""
        merged_vars = {**self.default_vars, **(variables or {})}
        return self.evaluator.evaluate(self.ast, merged_vars)
    
    def evaluate_array(self, x_values: List[float], 
                       var_name: str = 'x') -> List[float]:
        """Evaluate the expression over an array of values."""
        results = []
        for x in x_values:
            try:
                result = self.evaluate({var_name: x})
                results.append(result)
            except (ValueError, ZeroDivisionError, NameError):
                results.append(float('nan'))
        return results


def parse_expression(expression: str) -> CompiledExpression:
    """Convenience function to parse an expression string."""
    return CompiledExpression(expression)


if __name__ == "__main__":
    # Test the parser
    test_expressions = [
        "3 * x + sin(x)",
        "x^2 + 2*x + 1",
        "sin(x^2) + ln(x)",
        "sqrt(x^2 + y^2)",
        "pi * r^2",
        "tan(x)",
        "1 / (1 + x^2)",
    ]
    
    for expr_str in test_expressions:
        print(f"\nExpression: {expr_str}")
        try:
            expr = parse_expression(expr_str)
            print(f"  AST: {expr.ast}")
            
            # Test evaluation
            if 'y' not in expr_str and 'r' not in expr_str:
                test_val = 1.0
                result = expr.evaluate({'x': test_val})
                print(f"  f({test_val}) = {result}")
        except Exception as e:
            print(f"  Error: {e}")
